File: app/src/main/java/com/quran/labs/androidquran/ui/TranslationManagerActivity.java
package com.quran.labs.androidquran.ui;

import com.google.android.material.snackbar.Snackbar;

import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.SparseIntArray;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.quran.labs.androidquran.QuranApplication;
import com.quran.labs.androidquran.R;
import com.quran.labs.androidquran.dao.translation.Translation;
import com.quran.labs.androidquran.dao.translation.TranslationHeader;
import com.quran.labs.androidquran.dao.translation.TranslationItem;
import com.quran.labs.androidquran.dao.translation.TranslationItemDisplaySort;
import com.quran.labs.androidquran.dao.translation.TranslationRowData;
import com.quran.labs.androidquran.database.DatabaseHandler;
import com.quran.labs.androidquran.presenter.translation.TranslationManagerPresenter;
import com.quran.labs.androidquran.service.QuranDownloadService;
import com.quran.labs.androidquran.service.util.DefaultDownloadReceiver;
import com.quran.labs.androidquran.service.util.QuranDownloadNotifier;
import com.quran.labs.androidquran.service.util.ServiceIntentHelper;
import com.quran.labs.androidquran.ui.adapter.DownloadedItemActionListener;
import com.quran.labs.androidquran.ui.adapter.DownloadedMenuActionListener;
import com.quran.labs.androidquran.ui.adapter.TranslationsAdapter;
import com.quran.labs.androidquran.util.QuranFileUtils;
import com.quran.labs.androidquran.util.QuranSettings;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import javax.inject.Inject;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.view.ActionMode;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import io.reactivex.disposables.Disposable;
import timber.log.Timber;

public class TranslationManagerActivity extends QuranActionBarActivity
    implements DefaultDownloadReceiver.SimpleDownloadListener, DownloadedMenuActionListener {

  public static final String TRANSLATION_DOWNLOAD_KEY = "TRANSLATION_DOWNLOAD_KEY";
  private static final String UPGRADING_EXTENSION = ".old";

  private List<TranslationItem> allItems;
  private List<TranslationItem> currentSortedDownloads;
  private List<TranslationItem> originalSortedDownloads;

  private SparseIntArray translationPositions;

  private TranslationsAdapter adapter;
  private TranslationItem downloadingItem;
  private String databaseDirectory;
  private QuranSettings quranSettings;
  private DefaultDownloadReceiver downloadReceiver = null;

  private Disposable onClickDownloadDisposable;
  private Disposable onClickRemoveDisposable;
  private Disposable onClickRankUpDisposable;
  private Disposable onClickRankDownDisposable;

  private ActionMode actionMode;
  private TranslationSelectionListener selectionListener;
  private DownloadedItemActionListener downloadedItemActionListener;

  @Inject TranslationManagerPresenter presenter;
  @Inject QuranFileUtils quranFileUtils;

  SwipeRefreshLayout translationSwipeRefresh;
  RecyclerView translationRecycler;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    ((QuranApplication) getApplication()).getApplicationComponent().inject(this);
    setContentView(R.layout.translation_manager);
    translationSwipeRefresh = findViewById(R.id.translation_swipe_refresh);
    translationRecycler = findViewById(R.id.translation_recycler);

    RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
    translationRecycler.setLayoutManager(layoutManager);

    adapter = new TranslationsAdapter(this);
    translationRecycler.setAdapter(adapter);
    selectionListener = new TranslationSelectionListener(adapter);

    databaseDirectory = quranFileUtils.getQuranDatabaseDirectory(this);

    final ActionBar actionBar = getSupportActionBar();
    if (actionBar != null) {
      actionBar.setDisplayHomeAsUpEnabled(true);
      actionBar.setTitle(R.string.prefs_translations);
    }

    quranSettings = QuranSettings.getInstance(this);
    onClickDownloadDisposable = adapter.getOnClickDownloadSubject().subscribe(this::downloadItem);
    onClickRemoveDisposable = adapter.getOnClickRemoveSubject().subscribe(this::removeItem);
    onClickRankUpDisposable = adapter.getOnClickRankUpSubject().subscribe(this::rankUpItem);
    onClickRankDownDisposable = adapter.getOnClickRankDownSubject().subscribe(this::rankDownItem);

    translationSwipeRefresh.setOnRefreshListener(this::onRefresh);
    presenter.bind(this);
    translationSwipeRefresh.setRefreshing(true);
    presenter.getTranslationsList(false);
  }

  @Override
  public void onStop() {
    if (downloadReceiver != null) {
      downloadReceiver.setListener(null);
      LocalBroadcastManager.getInstance(this)
          .unregisterReceiver(downloadReceiver);
      downloadReceiver = null;
    }
    super.onStop();
  }

  @Override
  protected void onDestroy() {
    presenter.unbind(this);
    onClickDownloadDisposable.dispose();
    onClickRemoveDisposable.dispose();
    onClickRankUpDisposable.dispose();
    onClickRankDownDisposable.dispose();
    super.onDestroy();
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    if (item.getItemId() == android.R.id.home) {
      finish();
      return true;
    } else {
      return super.onOptionsItemSelected(item);
    }
  }

  @Override
  public void handleDownloadSuccess() {
    if (downloadingItem != null) {
      if (downloadingItem.exists()) {
        try {
          File f = new File(databaseDirectory,
              downloadingItem.getTranslation().getFileName() + UPGRADING_EXTENSION);
          if (f.exists()) {
            f.delete();
          }
        } catch (Exception e) {
          Timber.d(e, "error removing old database file");
        }
      }

      List<TranslationItem> sortedItems = sortedDownloadedItems();
      int lastDisplayOrder = sortedItems.isEmpty() ? 1 :
              sortedItems.get(sortedItems.size() - 1).getDisplayOrder();

      final Translation translation = downloadingItem.getTranslation();
      TranslationItem updated = new TranslationItem(translation,
              translation.getCurrentVersion(), lastDisplayOrder + 1);
      updateTranslationItem(updated);

      // update active translations and add this item to it
      QuranSettings settings = QuranSettings.getInstance(this);
      Set<String> activeTranslations = settings.getActiveTranslations();
      activeTranslations.add(downloadingItem.getTranslation().getFileName());
      settings.setActiveTranslations(activeTranslations);
    }
    downloadingItem = null;
    generateListItems();
  }

  @Override
  public void handleDownloadFailure(int errId) {
    if (downloadingItem != null && downloadingItem.exists()) {
      try {
        File f = new File(databaseDirectory,
            downloadingItem.getTranslation().getFileName() + UPGRADING_EXTENSION);
        File destFile = new File(databaseDirectory, downloadingItem.getTranslation().getFileName());
        if (f.exists() && !destFile.exists()) {
          f.renameTo(destFile);
        } else {
          f.delete();
        }
      } catch (Exception e) {
        Timber.d(e, "error restoring translation after failed download");
      }
    }
    downloadingItem = null;
  }

  private void onRefresh() {
    presenter.getTranslationsList(true);
  }

  private void updateTranslationItem(TranslationItem updated) {
    int id = updated.getTranslation().getId();
    int allItemsIndex = translationPositions.get(id);
    if (allItems != null && allItems.size() > allItemsIndex) {
      allItems.remove(allItemsIndex);
      allItems.add(allItemsIndex, updated);
    }
    presenter.updateItem(updated);
  }

  private void updateDownloadedItems() {
    final List<TranslationRowData> translations = adapter.getTranslations();
    final int downloadedItemCount = currentSortedDownloads.size();
    if (downloadedItemCount + 1 <= translations.size()) {
      for (int i = 0; i < downloadedItemCount; i++) {
        translations.remove(1);
      }

      translations.addAll(1, currentSortedDownloads);
      adapter.setTranslations(translations);
      adapter.notifyDataSetChanged();
    }
  }

  public void onErrorDownloadTranslations() {
    translationSwipeRefresh.setRefreshing(false);
    Snackbar
        .make(translationRecycler, R.string.error_getting_translation_list, Snackbar.LENGTH_SHORT)
        .show();
  }

  public void onTranslationsUpdated(List<TranslationItem> items) {
    translationSwipeRefresh.setRefreshing(false);
    SparseIntArray itemsSparseArray = new SparseIntArray(items.size());
    for (int i = 0, itemsSize = items.size(); i < itemsSize; i++) {
      TranslationItem item = items.get(i);
      itemsSparseArray.put(item.getTranslation().getId(), i);
    }
    allItems = items;
    translationPositions = itemsSparseArray;

    generateListItems();
  }

  private void generateListItems() {
    if (allItems == null) {
      return;
    }

    List<TranslationItem> downloaded = new ArrayList<>();
    List<TranslationItem> notDownloaded = new ArrayList<>();
    for (int i = 0, allItemsSize = allItems.size(); i < allItemsSize; i++) {
      TranslationItem item = allItems.get(i);
      if (item.exists()) {
        downloaded.add(item);
      } else {
        notDownloaded.add(item);
      }
    }

    List<TranslationRowData> result = new ArrayList<>();
    if (downloaded.size() > 0) {
      TranslationHeader hdr = new TranslationHeader(getString(R.string.downloaded_translations));
      result.add(hdr);

      // sort by display order
      Collections.sort(downloaded, new TranslationItemDisplaySort());

      boolean needsUpgrade = false;
      for (TranslationItem item : downloaded) {
        result.add(item);
        needsUpgrade = needsUpgrade || item.needsUpgrade();
      }

      if (!needsUpgrade) {
        quranSettings.setHaveUpdatedTranslations(false);
      }
    }
    originalSortedDownloads = new ArrayList<>(downloaded);
    currentSortedDownloads = new ArrayList<>(downloaded);

    result.add(new TranslationHeader(getString(R.string.available_translations)));

    result.addAll(notDownloaded);

    adapter.setTranslations(result);
    adapter.notifyDataSetChanged();
  }

  private void downloadItem(TranslationRowData translationRowData) {
    TranslationItem selectedItem = (TranslationItem) translationRowData;
    if (selectedItem.exists() && !selectedItem.needsUpgrade()) {
      return;
    }

    downloadingItem = selectedItem;

    final Translation translation = selectedItem.getTranslation();
    DatabaseHandler.clearDatabaseHandlerIfExists(translation.getFileName());
    if (downloadReceiver == null) {
      downloadReceiver = new DefaultDownloadReceiver(this,
          QuranDownloadService.DOWNLOAD_TYPE_TRANSLATION);
      LocalBroadcastManager.getInstance(this).registerReceiver(
          downloadReceiver, new IntentFilter(
              QuranDownloadNotifier.ProgressIntent.INTENT_NAME));
    }
    downloadReceiver.setListener(this);

    // actually start the download
    String url = translation.getFileUrl();
    String destination = databaseDirectory;
    Timber.d("downloading %s to %s", url, destination);

    if (selectedItem.exists()) {
      try {
        File f = new File(destination, translation.getFileName());
        if (f.exists()) {
          File newPath = new File(destination,
              translation.getFileName() + UPGRADING_EXTENSION);
          if (newPath.exists()) {
            newPath.delete();
          }
          f.renameTo(newPath);
        }
      } catch (Exception e) {
        Timber.d(e, "error backing database file up");
      }
    }

    // start the download
    String notificationTitle = selectedItem.name();
    Intent intent = ServiceIntentHelper.getDownloadIntent(this, url,
        destination, notificationTitle, TRANSLATION_DOWNLOAD_KEY,
        QuranDownloadService.DOWNLOAD_TYPE_TRANSLATION);
    String filename = selectedItem.getTranslation().getFileName();
    if (url.endsWith("zip")) {
      filename += ".zip";
    }
    intent.putExtra(QuranDownloadService.EXTRA_OUTPUT_FILE_NAME, filename);
    startService(intent);
  }

  private void removeItem(final TranslationRowData translationRowData) {
    if (adapter == null) {
      return;
    }

    final TranslationItem selectedItem =
        (TranslationItem) translationRowData;
    String msg = String.format(getString(R.string.remove_dlg_msg), selectedItem.name());
    AlertDialog.Builder builder = new AlertDialog.Builder(this);
    builder.setTitle(R.string.remove_dlg_title)
        .setMessage(msg)
        .setPositiveButton(R.string.remove_button,
            (dialog, id) -> {
              if (removeTranslation(selectedItem.getTranslation().getFileName())) {
                TranslationItem updatedItem = selectedItem.withTranslationRemoved();
                updateTranslationItem(updatedItem);

                // remove from active translations
                QuranSettings settings = QuranSettings.getInstance(this);
                Set<String> activeTranslations = settings.getActiveTranslations();
                activeTranslations.remove(selectedItem.getTranslation().getFileName());
                settings.setActiveTranslations(activeTranslations);
                generateListItems();
              }
            })
        .setNegativeButton(R.string.cancel,
            (dialog, i) -> dialog.dismiss());
    builder.show();
  }

  private List<TranslationItem> sortedDownloadedItems() {
    final ArrayList<TranslationItem> result = new ArrayList<>();
    for (TranslationItem item : allItems) {
      if (item.exists()) result.add(item);
    }
    Collections.sort(result, new TranslationItemDisplaySort());
    return result;
  }

  private void rankDownItem(TranslationRowData targetRow) {
    final TranslationItem targetItem = (TranslationItem) targetRow;
    final int targetTranslationId = targetItem.getTranslation().getId();

    int targetIndex = -1;
    for (int i = 0; i < currentSortedDownloads.size(); i++) {
      if (currentSortedDownloads.get(i).getTranslation().getId() == targetTranslationId) {
        targetIndex = i;
        break;
      }
    }

    if (targetIndex >= 0) {
      currentSortedDownloads.remove(targetIndex);
      final TranslationItem updatedItem =
              targetItem.withDisplayOrder(targetItem.getDisplayOrder() + 1);
      if (targetIndex + 1 < currentSortedDownloads.size()) {
        currentSortedDownloads.add(targetIndex + 1, updatedItem);
      } else {
        currentSortedDownloads.add(updatedItem);
      }
      updateDownloadedItems();
    }
  }

  private void rankUpItem(TranslationRowData targetRow) {
    final TranslationItem targetItem = (TranslationItem) targetRow;
    final int targetTranslationId = targetItem.getTranslation().getId();

    int targetIndex = -1;
    for (int i = 0; i < currentSortedDownloads.size(); i++) {
      if (currentSortedDownloads.get(i).getTranslation().getId() == targetTranslationId) {
        targetIndex = i;
        break;
      }
    }

    if (targetIndex >= 0) {
      currentSortedDownloads.remove(targetIndex);
      final TranslationItem updatedItem =
              targetItem.withDisplayOrder(targetItem.getDisplayOrder() - 1);
      currentSortedDownloads.add(Math.max(targetIndex - 1, 0), updatedItem);
      updateDownloadedItems();
    }
  }

  private void updateTranslationOrdersIfNecessary() {
    if (!originalSortedDownloads.equals(currentSortedDownloads)) {
      final List<TranslationItem> normalizedSortOrders = new ArrayList<>();
      for (int i = 0; i < currentSortedDownloads.size(); i++) {
        normalizedSortOrders.add(currentSortedDownloads.get(i).withDisplayOrder(i + 1));
      }
      originalSortedDownloads.clear();
      originalSortedDownloads.addAll(normalizedSortOrders);
      currentSortedDownloads.clear();
      currentSortedDownloads.addAll(normalizedSortOrders);
      presenter.updateItemOrdering(normalizedSortOrders);
    }
  }

  private boolean removeTranslation(String fileName) {
    String path = quranFileUtils.getQuranDatabaseDirectory(TranslationManagerActivity.this);
    if (path != null) {
      path += File.separator + fileName;
      File f = new File(path);
      return f.delete();
    }
    return false;
  }

  @Override
  public void startMenuAction(TranslationItem item, DownloadedItemActionListener aDownloadedItemActionListener) {
    downloadedItemActionListener = aDownloadedItemActionListener;
    if (actionMode != null) {
      actionMode.finish();
      selectionListener.clearSelection();
    } else {
      selectionListener.handleSelection(item);
      actionMode = startSupportActionMode(new ModeCallback());
    }
  }

  @Override
  public void finishMenuAction() {
    if (actionMode != null) {
      actionMode.finish();
    }
    selectionListener.clearSelection();
    downloadedItemActionListener = null;
  }

  class TranslationSelectionListener {
    private final TranslationsAdapter adapter;

    TranslationSelectionListener(TranslationsAdapter anAdapter) {
      adapter = anAdapter;
    }

    void handleSelection(TranslationItem item) {
      adapter.setSelectedItem(item);
    }

    void clearSelection() {
      adapter.setSelectedItem(null);
    }
  }

  private class ModeCallback implements ActionMode.Callback  {
    @Override
    public boolean onCreateActionMode(ActionMode mode, Menu menu) {
      MenuInflater inflater = getMenuInflater();
      inflater.inflate(R.menu.downloaded_translation_menu, menu);
      return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
      return false;
    }

    @Override
    public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
      switch(item.getItemId()) {
        case R.id.dtm_delete:
          if (downloadedItemActionListener != null) downloadedItemActionListener.handleDeleteItemAction();
          endAction();
          break;
        case R.id.dtm_move_up:
          if (downloadedItemActionListener != null) downloadedItemActionListener.handleRankUpItemAction();
          break;
        case R.id.dtm_move_down:
          if (downloadedItemActionListener != null) downloadedItemActionListener.handleRankDownItemAction();
          break;
      }
      return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode mode) {
      if (mode == actionMode) {
        selectionListener.clearSelection();
        actionMode = null;
        updateTranslationOrdersIfNecessary();
      }
    }

    private void endAction() {
      if (actionMode != null) {
        selectionListener.clearSelection();
        actionMode.finish();
      }
    }
  }
}

File: app/src/main/java/com/quran/labs/androidquran/ui/TranslationManagerActivity.kt
package com.quran.labs.androidquran.ui

import android.content.DialogInterface
import android.content.IntentFilter
import android.os.Bundle
import android.util.SparseIntArray
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.snackbar.Snackbar
import com.quran.labs.androidquran.QuranApplication
import com.quran.labs.androidquran.R
import com.quran.labs.androidquran.dao.translation.TranslationHeader
import com.quran.labs.androidquran.dao.translation.TranslationItem
import com.quran.labs.androidquran.dao.translation.TranslationRowData
import com.quran.labs.androidquran.database.DatabaseHandler.Companion.clearDatabaseHandlerIfExists
import com.quran.labs.androidquran.presenter.translation.TranslationManagerPresenter
import com.quran.labs.androidquran.service.QuranDownloadService
import com.quran.labs.androidquran.service.util.DefaultDownloadReceiver
import com.quran.labs.androidquran.service.util.DefaultDownloadReceiver.SimpleDownloadListener
import com.quran.labs.androidquran.service.util.QuranDownloadNotifier
import com.quran.labs.androidquran.service.util.ServiceIntentHelper.getDownloadIntent
import com.quran.labs.androidquran.ui.adapter.DownloadedItemActionListener
import com.quran.labs.androidquran.ui.adapter.DownloadedMenuActionListener
import com.quran.labs.androidquran.ui.adapter.TranslationsAdapter
import com.quran.labs.androidquran.util.QuranFileUtils
import com.quran.labs.androidquran.util.QuranSettings
import io.reactivex.rxjava3.disposables.Disposable
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import timber.log.Timber
import java.io.File
import javax.inject.Inject
import kotlin.math.max

class TranslationManagerActivity : AppCompatActivity(), SimpleDownloadListener,
  DownloadedMenuActionListener {
  private var allItems: List<TranslationItem> = emptyList()
  private var currentSortedDownloads: List<TranslationItem> = emptyList()
  private var originalSortedDownloads: List<TranslationItem> = emptyList()
  private var translationPositions: SparseIntArray = SparseIntArray()
  private var downloadingItem: TranslationItem? = null
  private var databaseDirectory: String? = null
  private var downloadReceiver: DefaultDownloadReceiver? = null
  private var actionMode: ActionMode? = null
  private var downloadedItemActionListener: DownloadedItemActionListener? = null

  @Inject
  lateinit var presenter: TranslationManagerPresenter

  @Inject
  lateinit var quranFileUtils: QuranFileUtils

  @Inject
  lateinit var quranSettings: QuranSettings

  private lateinit var adapter: TranslationsAdapter
  private lateinit var selectionListener: TranslationSelectionListener
  private lateinit var onClickDownloadDisposable: Disposable
  private lateinit var onClickRemoveDisposable: Disposable
  private lateinit var onClickRankUpDisposable: Disposable
  private lateinit var onClickRankDownDisposable: Disposable

  private lateinit var translationSwipeRefresh: SwipeRefreshLayout
  private lateinit var translationRecycler: RecyclerView

  private val scope = MainScope()

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    (application as QuranApplication).applicationComponent.inject(this)
    setContentView(R.layout.translation_manager)
    translationSwipeRefresh = findViewById(R.id.translation_swipe_refresh)
    translationRecycler = findViewById(R.id.translation_recycler)
    val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
    translationRecycler.setLayoutManager(layoutManager)
    adapter = TranslationsAdapter(this)
    translationRecycler.setAdapter(adapter)
    selectionListener = TranslationSelectionListener(adapter)
    databaseDirectory = quranFileUtils.getQuranDatabaseDirectory(this)
    val actionBar = supportActionBar
    if (actionBar != null) {
      actionBar.setDisplayHomeAsUpEnabled(true)
      actionBar.setTitle(R.string.prefs_translations)
    }
    onClickDownloadDisposable = adapter.getOnClickDownloadSubject()
      .subscribe { translationRowData: TranslationRowData -> downloadItem(translationRowData) }
    onClickRemoveDisposable = adapter.getOnClickRemoveSubject()
      .subscribe { translationRowData: TranslationRowData -> removeItem(translationRowData) }
    onClickRankUpDisposable = adapter.getOnClickRankUpSubject()
      .subscribe { targetRow: TranslationRowData -> rankUpItem(targetRow) }
    onClickRankDownDisposable = adapter.getOnClickRankDownSubject()
      .subscribe { targetRow: TranslationRowData -> rankDownItem(targetRow) }
    translationSwipeRefresh.setOnRefreshListener { onRefresh() }
    translationSwipeRefresh.isRefreshing = true
    refreshTranslations()
  }

  public override fun onStop() {
    val receiver = downloadReceiver
    if (receiver != null) {
      receiver.setListener(null)
      LocalBroadcastManager.getInstance(this)
        .unregisterReceiver(receiver)
      downloadReceiver = null
    }
    super.onStop()
  }

  override fun onDestroy() {
    scope.cancel()
    onClickDownloadDisposable.dispose()
    onClickRemoveDisposable.dispose()
    onClickRankUpDisposable.dispose()
    onClickRankDownDisposable.dispose()
    super.onDestroy()
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean {
    return if (item.itemId == android.R.id.home) {
      finish()
      true
    } else {
      super.onOptionsItemSelected(item)
    }
  }

  override fun handleDownloadSuccess() {
    val downloadingItem = downloadingItem
    if (downloadingItem != null) {
      if (downloadingItem.exists()) {
        try {
          val f = File(
            databaseDirectory,
            downloadingItem.translation.fileName + UPGRADING_EXTENSION
          )
          if (f.exists()) {
            f.delete()
          }
        } catch (e: Exception) {
          Timber.d(e, "error removing old database file")
        }
      }

      // TODO: we can avoid the cost of sorting once we can listen to db updates
      // in which case we'd set the local version as -1 so it gets properly assigned after.
      val sortedItems = sortedDownloadedItems()
      val lastDisplayOrder =
        if (sortedItems.isEmpty()) 0 else sortedItems[sortedItems.size - 1].displayOrder
      val (_, _, currentVersion) = downloadingItem.translation
      updateTranslationItem(
        downloadingItem.withLocalVersionAndDisplayOrder(
          currentVersion, lastDisplayOrder + 1
        )
      )

      // update active translations and add this item to it
      val settings = QuranSettings.getInstance(this)
      val activeTranslations = settings.activeTranslations
      activeTranslations.add(downloadingItem.translation.fileName)
      settings.activeTranslations = activeTranslations
    }
    this.downloadingItem = null
    generateListItems()
  }

  override fun handleDownloadFailure(errId: Int) {
    val downloadingItem = downloadingItem
    if (downloadingItem != null && downloadingItem.exists()) {
      try {
        val f = File(
          databaseDirectory,
          downloadingItem.translation.fileName + UPGRADING_EXTENSION
        )
        val destFile = File(databaseDirectory, downloadingItem.translation.fileName)
        if (f.exists() && !destFile.exists()) {
          f.renameTo(destFile)
        } else {
          f.delete()
        }
      } catch (e: Exception) {
        Timber.d(e, "error restoring translation after failed download")
      }
    }
    this.downloadingItem = null
  }

  private fun onRefresh() {
    refreshTranslations(true)
  }

  private fun refreshTranslations(forceDownload: Boolean = false) {
    presenter.getTranslations(forceDownload)
      .onEach { onTranslationsUpdated(it) }
      .catch { onErrorDownloadTranslations() }
      .launchIn(scope)
  }

  private fun updateTranslationItem(updated: TranslationItem) {
    val id = updated.translation.id
    val allItemsIndex = translationPositions[id]
    if (allItems.size > allItemsIndex) {
      allItems = allItems.toMutableList().apply {
        removeAt(allItemsIndex)
        add(allItemsIndex, updated)
      }
    }
    scope.launch {
      presenter.updateItem(updated)
    }
  }

  private fun updateDownloadedItems() {
    val translations = adapter.getTranslations().toMutableList()
    val downloadedItemCount = currentSortedDownloads.size
    if (downloadedItemCount + 1 <= translations.size) {
      for (i in 0 until downloadedItemCount) {
        translations.removeAt(1)
      }
      translations.addAll(1, currentSortedDownloads)
      adapter.setTranslations(translations)
      adapter.notifyDataSetChanged()
    }
  }

  private fun onErrorDownloadTranslations() {
    translationSwipeRefresh.isRefreshing = false
    Snackbar
      .make(
        translationRecycler,
        R.string.error_getting_translation_list,
        Snackbar.LENGTH_SHORT
      )
      .show()
  }

  private fun onTranslationsUpdated(items: List<TranslationItem>) {
    translationSwipeRefresh.isRefreshing = false
    val itemsSparseArray = SparseIntArray(items.size)
    var i = 0
    val itemsSize = items.size
    while (i < itemsSize) {
      val (translation) = items[i]
      itemsSparseArray.put(translation.id, i)
      i++
    }
    allItems = items
    translationPositions = itemsSparseArray
    generateListItems()
  }

  private fun generateListItems() {
    val (downloaded, notDownloaded) = allItems.partition { it.exists() }

    // sort by display order
    val sortedDownloads = downloaded.sortedBy { it.displayOrder }

    val resultList = buildList {
      if (downloaded.isNotEmpty()) {
        add(TranslationHeader(getString(R.string.downloaded_translations)))
        addAll(sortedDownloads)
      }
      add(TranslationHeader(getString(R.string.available_translations)))
      addAll(notDownloaded)
    }

    val needsUpgrade = sortedDownloads.any { it.needsUpgrade() }
    if (!needsUpgrade) {
      quranSettings.setHaveUpdatedTranslations(false)
    }

    originalSortedDownloads = ArrayList(downloaded)
    currentSortedDownloads = ArrayList(downloaded)
    adapter.setTranslations(resultList)
    adapter.notifyDataSetChanged()
  }

  private fun downloadItem(translationRowData: TranslationRowData) {
    val selectedItem = translationRowData as TranslationItem
    if (selectedItem.exists() && !selectedItem.needsUpgrade()) {
      return
    }
    downloadingItem = selectedItem
    val (_, _, _, _, _, fileName, url) = selectedItem.translation
    clearDatabaseHandlerIfExists(fileName)
    if (downloadReceiver == null) {
      val downloadReceiver = DefaultDownloadReceiver(
        this,
        QuranDownloadService.DOWNLOAD_TYPE_TRANSLATION
      )
      LocalBroadcastManager.getInstance(this).registerReceiver(
        downloadReceiver, IntentFilter(
          QuranDownloadNotifier.ProgressIntent.INTENT_NAME
        )
      )
      this.downloadReceiver = downloadReceiver
    }
    downloadReceiver!!.setListener(this)

    // actually start the download
    val destination = databaseDirectory
    Timber.d("downloading %s to %s", url, destination)
    if (selectedItem.exists()) {
      try {
        val f = File(destination, fileName)
        if (f.exists()) {
          val newPath = File(
            destination,
            fileName + UPGRADING_EXTENSION
          )
          if (newPath.exists()) {
            newPath.delete()
          }
          f.renameTo(newPath)
        }
      } catch (e: Exception) {
        Timber.d(e, "error backing database file up")
      }
    }

    // start the download
    val notificationTitle = selectedItem.name()
    val intent = getDownloadIntent(
      this, url,
      destination, notificationTitle, TRANSLATION_DOWNLOAD_KEY,
      QuranDownloadService.DOWNLOAD_TYPE_TRANSLATION
    )
    var filename = selectedItem.translation.fileName
    if (url.endsWith("zip")) {
      filename += ".zip"
    }
    intent.putExtra(QuranDownloadService.EXTRA_OUTPUT_FILE_NAME, filename)
    startService(intent)
  }

  private fun removeItem(translationRowData: TranslationRowData) {
    val selectedItem = translationRowData as TranslationItem
    val msg = String.format(getString(R.string.remove_dlg_msg), selectedItem.name())
    val builder = AlertDialog.Builder(this)
    builder.setTitle(R.string.remove_dlg_title)
      .setMessage(msg)
      .setPositiveButton(
        com.quran.mobile.common.ui.core.R.string.remove_button
      ) { _: DialogInterface?, _: Int ->
        if (removeTranslation(selectedItem.translation.fileName)) {
          val updatedItem = selectedItem.withTranslationRemoved()
          updateTranslationItem(updatedItem)

          // remove from active translations
          val settings = QuranSettings.getInstance(this)
          val activeTranslations = settings.activeTranslations
          activeTranslations.remove(selectedItem.translation.fileName)
          settings.activeTranslations = activeTranslations
          generateListItems()
        }
      }
      .setNegativeButton(
        com.quran.mobile.common.ui.core.R.string.cancel
      ) { dialog: DialogInterface, i: Int -> dialog.dismiss() }
    builder.show()
  }

  private fun sortedDownloadedItems(): List<TranslationItem> {
    return allItems.filter { it.exists() }.sortedBy { it.displayOrder }
  }

  private fun rankDownItem(targetRow: TranslationRowData) {
    val targetItem = targetRow as TranslationItem
    val targetTranslationId = targetItem.translation.id
    val targetIndex = currentSortedDownloads.indexOfFirst { it.translation.id == targetTranslationId }
    if (targetIndex >= 0) {
      val sortedDownloads = currentSortedDownloads.toMutableList()
      sortedDownloads.removeAt(targetIndex)
      val updatedItem = targetItem.withDisplayOrder(targetItem.displayOrder + 1)
      if (targetIndex + 1 < sortedDownloads.size) {
        sortedDownloads.add(targetIndex + 1, updatedItem)
      } else {
        sortedDownloads.add(updatedItem)
      }
      currentSortedDownloads = sortedDownloads
      updateDownloadedItems()
    }
  }

  private fun rankUpItem(targetRow: TranslationRowData) {
    val targetItem = targetRow as TranslationItem
    val targetTranslationId = targetItem.translation.id
    val targetIndex = currentSortedDownloads.indexOfFirst { it.translation.id == targetTranslationId }
    if (targetIndex >= 0) {
      val sortedDownloads = currentSortedDownloads.toMutableList()
      sortedDownloads.removeAt(targetIndex)
      val updatedItem = targetItem.withDisplayOrder(targetItem.displayOrder - 1)
      sortedDownloads.add(max(targetIndex - 1, 0), updatedItem)
      currentSortedDownloads = sortedDownloads
      updateDownloadedItems()
    }
  }

  private fun updateTranslationOrdersIfNecessary() {
    if (originalSortedDownloads != currentSortedDownloads) {
      val normalizedSortOrders: List<TranslationItem> =
        currentSortedDownloads.mapIndexed { index, item ->
          item.withDisplayOrder(index + 1)
        }

      originalSortedDownloads = normalizedSortOrders
      currentSortedDownloads = normalizedSortOrders
      scope.launch {
        presenter.updateItemOrdering(normalizedSortOrders)
      }
    }
  }

  private fun removeTranslation(fileName: String): Boolean {
    var path = quranFileUtils.getQuranDatabaseDirectory(this@TranslationManagerActivity)
    if (path != null) {
      path += File.separator + fileName
      val f = File(path)
      return f.delete()
    }
    return false
  }

  override fun startMenuAction(
    item: TranslationItem,
    downloadedItemActionListener: DownloadedItemActionListener?
  ) {
    this.downloadedItemActionListener = downloadedItemActionListener
    if (actionMode != null) {
      actionMode!!.finish()
      selectionListener.clearSelection()
    } else {
      selectionListener.handleSelection(item)
      actionMode = startSupportActionMode(ModeCallback())
    }
  }

  override fun finishMenuAction() {
    actionMode?.finish()
    selectionListener.clearSelection()
    downloadedItemActionListener = null
  }

  internal class TranslationSelectionListener(private val adapter: TranslationsAdapter) {
    fun handleSelection(item: TranslationItem?) {
      adapter.setSelectedItem(item)
    }

    fun clearSelection() {
      adapter.setSelectedItem(null)
    }
  }

  private inner class ModeCallback : ActionMode.Callback {
    override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
      val inflater = menuInflater
      inflater.inflate(R.menu.downloaded_translation_menu, menu)
      return true
    }

    override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
      return false
    }

    override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
      val itemId = item.itemId
      when (itemId) {
          R.id.dtm_delete -> {
            downloadedItemActionListener?.handleDeleteItemAction()
            endAction()
          }
          R.id.dtm_move_up -> {
            downloadedItemActionListener?.handleRankUpItemAction()
          }
          R.id.dtm_move_down -> {
            downloadedItemActionListener?.handleRankDownItemAction()
          }
      }
      return false
    }

    override fun onDestroyActionMode(mode: ActionMode) {
      if (mode === actionMode) {
        selectionListener.clearSelection()
        actionMode = null
        updateTranslationOrdersIfNecessary()
      }
    }

    private fun endAction() {
      if (actionMode != null) {
        selectionListener.clearSelection()
        actionMode!!.finish()
      }
    }
  }

  companion object {
    const val TRANSLATION_DOWNLOAD_KEY = "TRANSLATION_DOWNLOAD_KEY"
    private const val UPGRADING_EXTENSION = ".old"
  }
}
